File: python/guidos-gorgeous-lasagna/lasagna.py
"""Functions used in preparing Guido's gorgeous lasagna.

Learn about Guido, the creator of the Python language:
https://en.wikipedia.org/wiki/Guido_van_Rossum

This is a module docstring, used to describe the functionality
of a module and its functions and/or classes.
"""


EXPECTED_BAKE_TIME = 40

def bake_time_remaining(elapsed_bake_time):
    """Calculate the bake time remaining.

    :param elapsed_bake_time: int - baking time already elapsed.
    :return: int - remaining bake time (in minutes) derived from 'EXPECTED_BAKE_TIME'.

    Function that takes the actual minutes the lasagna has been in the oven as
    an argument and returns how many minutes the lasagna still needs to bake
    based on the `EXPECTED_BAKE_TIME`.
    """

    return EXPECTED_BAKE_TIME - elapsed_bake_time


def preparation_time_in_minutes(number_of_layers):
    """Calculate the preparation time based on number of layers of your lasagna.

    :param number_of_layers: int - number of layers of a lasagna.
    :return: int - expected preparation time in minutes.
    """
    return number_of_layers * 2


def elapsed_time_in_minutes(number_of_layers, elapsed_bake_time):
    """Calculate the elapsed cooking time of your lasagna.

    :param number_of_layers: int - number of layers of a lasagna.
    :param elapsed_bake_time: int - the time in minutes the same lasagna has been on oven.
    :return: int - time elapsed in minutes.
    """
    return preparation_time_in_minutes(number_of_layers) + elapsed_bake_time

File: solutions/elixir/lasagna/1/lib/lasagna.ex
defmodule Lasagna do
  
  @doc """
  Returns how many minutes the lasagna should be in the oven.
  """
  def expected_minutes_in_oven(), do: 40


  @doc """
  Takes the actual minutes the lasagna has been in the oven as 
  a argument and returns how many minutes the lasagna still has 
  to remain in the oven
  """
  def remaining_minutes_in_oven(actual_minutes) do
    expected_minutes_in_oven() - actual_minutes
  end


  @doc """
  Takes the number of layers you added to the lasagna as a argument
  and returns how many minutes you spent preparing the lasagna
  """
  def preparation_time_in_minutes(number_of_layers) do
    number_of_layers * 2
  end


  @doc """
  Takes two arguments: the first argument is the number of layers you added 
  to the lasagna, and the second argument is the number of minutes the lasagna 
  has been in the oven. 
  Returns how many minutes in total you've worked on cooking the lasagna, 
  which is the sum of the preparation time in minutes, and the time in minutes 
  the lasagna has spent in the oven at the moment.
  """
  def total_time_in_minutes(number_of_layers, minutes_in_oven) do
    preparation_time_in_minutes(number_of_layers) + minutes_in_oven
  end


  @doc """
  Returns a message indicating that the lasagna is ready to eat.
  """
  def alarm() do
    "Ding!"
  end

end
